keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,359 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Starting with Kubernetes 1.22, a new `CPUManager` flag has facilitated the use of `CPUManager` Policy options(#2625) which enable users to customize their behavior based on workload requirements without having to introduce an entirely new policy.
+These policy options work together to ensure an optimized cpu set is allocated for workloads running on a cluster.
+The three policy options that already exist are `full-pcpus-only`(#2625) and `distribute-cpus-across-numa` (#2902) and `align-by-socket` (#3327).
+With this KEP, a new `CPUManager` policy option is introduced which ensures that reservedSystemCPUs are reserved for system daemons or interrupt processing and are not used by burstable and best-effort pods.
+
+
+## Motivation
+
+The static policy is used to reduce latency or improve performance. If you want to move system daemons or interrupt processing to dedicated cores, the oblivious way is use the reservedSystemCPUs option. But in current implementation this isolation is implemented only for guaranteed pods not for burstable and best-effort pods.
+Admission is only comparing the cpu requests against the allocatable cpus. Since the cpu limit are higher than the request, it allows burstable and best-effort pods to use up the capacity of reservedSystemCPUs option and cause the OS/Systemd services to starve in real life deployments.
+
+### Goals
+ * Ensure the reservedSystemCPUs are only used by system daemons or interrupt processing not by pods.
+
+### Non-Goals
+
+## Proposal
+
+We propose to add a new `CPUManager` policy option called `strict-cpu-reservation` to the `static` policy of `CPUManager`.
+With this policy option, we remove the reserved cores from the list of all available cores at the stage of calculation DefaultCPUSet. As a result, burstable and best-effort containers are launched with a cpuset in which the reserved cores are excluded. 
+
+### Risks and Mitigations
+
+The risks of adding this new feature are quite low.
+It is isolated to a specific policy option within the `CPUManager`, and is protected both by the option itself, as well as the `CPUManagerPolicyAlphaOptions` feature gate (which is disabled by default).
+
```

Comment:
Currently with the presence of `reservedSystemCPUs` under `defaultCPUset` we have the ability to accommodate best effort and burstable pods on a node even if it is resource constrained (e.g. Guaranteed pods occupying majority of CPUs exclusively). But with this option enabled, we could end up in a lot of evictions.

The eviction manager on the node prioritizes pods based on their QoS Class. We should be okay in case of standalone deployments as the pods could go other nodes that have enough resources but what worries me is the fact that there are a lot of infrastructure components (e.g. Device plugins) deployed as daemon sets that might not necessarily belong to Guaranteed QoS class (I took a quick look at SR-IOV device plugin: https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/blob/v3.6.2/deployments/sriovdp-daemonset.yaml#L49-L54,  we can perhaps find other such examples). We could end up in a never ending cycle of eviction by Kubelet eviction manager and scheduler reschedulng the pods on the node by the daemon set controller.

What is our mitigation strategy for this ?

Reply:
High performance workloads, the typical type that using SRIOV/DPDK, are also typically deployed as Guaranteed QoS, they are not affected when strict-cpu-reservation option is enabled.
You enable 'strict-cpu-reservation' when you have to protect infrastructure services, all the systemd daemons and interrupt processing, from busty workloads. Telco is a typical use case where we do our business. If you don't have this issue, then don't enable 'strict-cpu-reservation' option.

Reply:
Will add feature interaction test cases with the existing policy options.

Reply:
> We explored a very similar concept in the not-so distant past and exhausting the burstable pool is something which I believe should not be done lightly. In this case at least we have an option (vs a builtin default behavior) which is a step in the right direction, but I still believe we need some form of protection to prevent exhaustion of the shared cpu pool. We will perhaps need a new option or to overload the semantics of existing options. 

Exactly my point!

I understand that the intention here is to protect infrastructure services (systemd daemons and interrupt processing) but my comment above was about the potential impact on infrastructure pods deployed using daemonset which might not always belong to Gu QoS Class especially when they are running on a node that has resource contention. Kubernetes doesn't have a way to distinguish between application pods and infrastructure pods. 
 
If we don't want to restrict `reservedSystemCPUs` for OS processes / systemd services, you can simply _not_ opt-in to this option but daemonsets are often used for deploying infrastructure pods on all the nodes irrespective of the configured Kubelet policy or configured option.  

Typically in telco use cases, the goal is to maximize CPU utilization by providing CPUs for exclusive allocation to high performance workloads. I would like us to consider and capture in the KEP what happens to BE and BU pods when high performance workloads have been allocated all the CPUs available (total -reserverd) to them? With the current proposal it sounds like we could we end up with an empty defaultCPUSet which IMO should be avoided. 

Does CPU exhaustion on a node result in eviction of such pods (even if they are important infra pods)? Could we end up in a never ending cycle of eviction by Kubelet eviction manager and scheduler reschedulng the pods respawned on the node by the daemon set controller?

Reply:
Infrastructure pods are, and should be, deployed with priorityClassName set to system-node-critical.
A bursty workload should compete with other bursty workload but not with the system daemons and interrupt processing which serves ALL processes running on the node. The 'strict-cpu-reservation' option is to provide this capability.
Mainstream hardware comes with increased number of CPU cores per Node. Someone reserves an unreasonable amount of CPU cores is a solution issue not a software issue.

Reply:
hey all, sorry for interjecting but this PR is something we are missing from Kubernetes for soooooo long it would be a shame to not see it merged :) so first I would like to thank Jing for working on this, and would like to emphasize how great and useful this would be for both our existing, and future RAN and packet core deployments.

this current limitation is often the (last) blocker of being able to place mission critical workloads on combined master+worker(+HCI storage) nodes, which results in an increased dimensioning for our customers. needing to make a hard choice between optimal sizing and guaranteed reliability leaves us in a tough spot especially when it comes to far edge deployments where shelf space is at a premium. I guess needless to say we expect this use-case to increase exponentially over the course of the next couple of years making this PR even more high-prio

I do understand the objections however, changing the base assumptions of working systems can (and probably will :) ) result in user errors, misconfiguration, and ultimately frustration. I do wonder if we could address these concerns with soft measures though?
the functionality is already opt-in, and as Jing mentioned it is a reasonable to expect expert mission critical workload administrators to already have multiple layers of protections in place. I think expanding the documentation with the appropriate details calling out the edge cases, providing guidance on when to and not to use the feature, and what are these additional layers of protections one can utilize to mitigate the described worst case scenarios should kind of do the trick. and probably a blog post, blog posts are always nice :)

in case you feel hard measures are required how about retaining the previous functionality and marrying it to the measure Jing proposed? if the main concern is making absolutely sure infra Pods are never preempted because of system reservation concerns, we could implement this feature in a way that if a Pod has a certain (maybe configurable? maybe "system-critical"?) priority classification we allow them, but only them to be placed on system reserved, while at the same time we ensure all the other shared/burstable Pods are still kept away from this set?
this functionality would be something which we might want naturally utilize anyway for some use-cases, it would serve as a guaranteed safeguard while also promote and reward "good" system administrator practices by "forcing" the user to designate mission critical Pods, a task which should be done anyway irrespective of this change. win-win-win in my book!

looking forward to your feedback, and fingers crossed, can't wait for this to be officially supported!

Reply:
Agreed on "expanding the documentation with the appropriate details calling out the edge cases, providing guidance on when to and not to use the feature, and what are these additional layers of protections one can utilize to mitigate the described worst case scenarios", good points, will update the KEP to provide input for this. In terms of algorithms, simple is golden, simple is king, I value the current implementation ported from previous attempt very much.

Reply:
@swatisehgal I mean I'm historically very much a proponent of having [3 distinct pools ](https://github.com/nokia/CPU-Pooler?tab=readme-ov-file#cpu-pools) just thought we don't want to overcomplicate things upstream ;)
I'm all for it if you all agree this is the way to go! be advised though, from my experience of deploying something like this in production a possible downside can be increased fragmentation. more fine grained pooling coupled with rigid pool allocations means we need to keep node/host group specs identical in case we want to have reserves in the cluster (and we usually do, upgrade/maintenance/spare etc.). so you either define the "same" pools on all your host groups across the cluster -possibly unused capacity on the nodes belonging to the host groups which currently do not host these "special workloads"-, or you only create 1+1 special host groups with the special pool config -"special" workloads become rigid, they can only be scheduled to these special host groups at any given point in time **and** you also kinda need to make sure everybody else is prioritized away from these special groups if there is spare capacity elsehwere-.

it's kind of a lose-lose situation, works for special purpose clusters but tough to manage in bigger/generic clusters. so for this reason I was always on the opinion [dynamic pooling](https://static.sched.com/hosted_files/kccncna19/f2/Kale-Csatari-Is%20There%20a%20Place%20for%20Performance%20Sensitive%20Workloads%20in%20Kubernetes%20V2.pdf) is the way to go eventually (slide 13). create some pools to begin with but have the ability to dynamically move cores between pools, on-demand, based on some algorithm

it sounded futuristic 5 years ago but maybe this is the time? :) however, it would definitely require larger scale changes in CPU Manager even today imho, so who knows. if we are thinking simpler in my mind multiple static pools is definitely a more precise solution however if your aim is to solve possible user misconfiguration it may not be the way to go: it exchanges one problem for another by simply punting it from node to cluster level (i.e. frustration will now come from fragmentation instead of node exhaustion).
if we like the system-critical on-reserve idea and but still think an additional layer of protection is required a policy-based approach might be simpler. Thinking of an "X% of capacity must be reserved for "shared"" kinda configuration (maybe per NUMA node). when a scheduling request comes into with an exclusive CPU request **and** the allocation would result in going below the limit (limits? could also have 2, one soft one hard) CPU Manager could simply deny the scheduling. this approach would 100% ensure desired dynamic capacity is always available on the node without statically cutting up the total CPU capacity in advance

Reply:
on second read I realized you also didn't propose a full static partitioning, my bad :)
so in that sense the "semi-flexible" pool and the policy-based approach would be kinda similar. just a matter of which config interface we think is better for the user, the logic behind it would be identical-ish

sounds like a good way to go about it!

Reply:
So, the suggestion is to introduce minSharedCPUs which is exclusive to reservedSystemCPUs and Excluively Allocated, and can be illustrated as following:
Total CPU cores (80) = reservedSystemCPUs (8) + Exclusively Allocated (20) + minSharedCPUs (12) + Shared (40)
Shared (40): Available for future exclusive allocation.

I am for it.

More comments / suggestions?

Reply:
Yes, pretty much. I would add that based on your example above at any given time `defaultCPUSet` would be composed of CPUs in "minSharedCPUs + Shared".

There is also one more consideration from scheduling point of view:
Total capacity = 80 
Allocatable = 72 (total - reserved) 
With 20 CPUs already exclusively allocated and assuming there are not a lot of burstable and non-exclusive GU pods on the node, the scheduler will consider such a node suitable for a workload that is requesting say 50 CPUs exclusively (given that there are 52 CPUs remaining on the node) but any request for exclusive CPU more than 40 cannot be fulfilled on the node due to configured `minSharedCPUs` and hence should result in admission failure IOW we would have another case of runway pod creation similar to https://github.com/kubernetes/kubernetes/issues/84869. 

Users might not expect this. Therefore this behavior should be highlighted in the KEP as well as documentation.

Reply:
> Users might not expect this. Therefore this behavior should be highlighted in the KEP as well as documentation.

factoring this into the admission code could be a beta or GA goal? I guess the worst case scenario from a Node API PoV would be adding (and maintaining) Shared to Allocatable, but it wouldn't require a total overhaul to solve like was the case with the topology use-case

do we still want to optionally allow system-critical Pods on reservedCPUs? I think the functionality is useful and orthogonal to this solution, could be another beta/GA goal maybe?

Reply:
> Not sure I understand what you are suggesting here because Allocatable (Total - reserved) is static during kubelet's lifecycle and doesn't correspond to actually available resources i.e.`Shared` would always be counted under Allocatable. Can you please elaborate?

my bad, wanted to reference "minSharedCPUs" instead of "Shared"

so the idea would be that the only thing the node scheduler needs to do in addition to tracking the real-time allocations on the node (which is already being done) is how much of the seemingly available capacity is not really available for the specific purpose of exclusive reservations. which is yes, (mostly) a static number.
this static info could be also captured as a node label, or the scheduler could read the configuration directly and do the calculation on its own for every admittance check etc, but idea here is that you don't need to change real-time resource tracking to defend against the edge case you described.

i.e. "ok I have an exclusive request for 50 CPUs, ok this node has 52 available nice, but wait a minute 50-minSharedCPUs is smaller than 50 so actually no that won't do"
i.e.2 no need to continuously track and update "Shared", which would make the implementation less impactful -> easier to do -> let's include it in the KEP :)

I only brought it up because you mentioned "this behavior should be highlighted in the KEP and the documentation", thought you didn't necessarily want to include the scheduling code in the KEP so I wanted to argue it is not actually that hard and let's do it! but I think we are on the same page, you clarified you even consider it as a mandatory part of the initial implementation rather than just a stretch goal.
awesome, I'm all for it!

Reply:
ok, this is my take of the discussion:

(1) From the scheduler point of view, there is a single set of CPUs to schedule pods, which is the node allocatable = total - reserved. It is at the node level, burstable and best-effort pods are allowed to run on the reserved cores. Yes, it is very common people/(our customers) are surprised that burstable and best-effort pods are allowed to run on the reserved cores since it is against "common" sense. The 'strict-cpu-reservation" feature is effectively removing this discrepancy.

(2) However, there is this concern of no longer be able to "hide" burstable and best-effort pods on the reserved cores, so we will add minSharedCPUs. Introducing minSharedCPUs will impact scheduler because it is a discrepancy that has to be taken care of. This scheduler impact changes this 'strict-cpu-reservation' feature from trivial to at least not trivial.

Please review and correct my understanding.

And, is it acceptable to document the scheduler impact in the alpha version while only do the scheduler code change in beta version? for the sake of simpler feature interaction and thanks to feature flag is by default false in the alpha version.

Reply:
I give a 2nd thought on the need of minSharedCPUs which is to deal with increased pod eviction concern raised by @swatisehgal and @ffromani.
> Memory pressure will trigger pod evocation, for which we define eviction threshold.
> CPU pressure results in CPU throttling, it does not trigger pod eviction.

What real issue is for this minSharedCPUs to resolve?

Would it be more correct to keep the 'strict-cpu-reservation' feature flag to false even under beta and stable? By that we remove any concern for the use cases people are not interested in the 'feature and do not enable the feature?

Reply:
@swatisehgal  @ffromani have you managed to checkout my questions?

Reply:
> Would it be more correct to keep the 'strict-cpu-reservation' feature flag to false even under beta and stable? By that we remove any concern for the use cases people are not interested in the 'feature and do not enable the feature?

I think the feature should be default to off, and, thruth to be told, I was working under the assumption this would be the case also for the future, in GA stage.

On top of the flag proper, there's the feature gate governing the avaialiblity of the flag itself.
When the feature gate is disabled (alpha), attempting to set the flag will lead to error - the flag is not available in the first place.
When the feature gate is enabled (beta and onwards), the flag can be set.

I'll comment about the rest ASAP.

Reply:
> ok, this is my take of the discussion:
> 
> (1) From the scheduler point of view, there is a single set of CPUs to schedule pods, which is the node allocatable = total - reserved. It is at the node level, burstable and best-effort pods are allowed to run on the reserved cores. Yes, it is very common people/(our customers) are surprised that burstable and best-effort pods are allowed to run on the reserved cores since it is against "common" sense. The 'strict-cpu-reservation" feature is effectively removing this discrepancy.
> 
> (2) However, there is this concern of no longer be able to "hide" burstable and best-effort pods on the reserved cores, so we will add minSharedCPUs. Introducing minSharedCPUs will impact scheduler because it is a discrepancy that has to be taken care of. This scheduler impact changes this 'strict-cpu-reservation' feature from trivial to at least not trivial.
> 
> Please review and correct my understanding.

Your understanding is accurate, thanks for summarizing  this.

> And, is it acceptable to document the scheduler impact in the alpha version while only do the scheduler code change in beta version? for the sake of simpler feature interaction and thanks to feature flag is by default false in the alpha version

My take would be that correct scheduling  behavior is a very critical aspect of this feature and needs to be handled at alpha stage itself. We need the right end to end behavior in case the feature is enabled at alpha stage. 

SIG scheduling's opinion should be weighed in as well on this subject.

Reply:
Hi @ffromani thank you for replying and confirming it is a valid option to keep the feature off forever (Alpha/Beta/GA).

Reply:
HI @swatisehgal, thank you for replying and confirming we are aligned on the feature has no eviction impact.
What other feature impact do you think could involve scheduler? At this moment, I don't see obvious scheduler impact, so I do not have any in the KEP itself.  How do I invite scheduler maintainers for comments? I did "/add-sig scheduling". Please advise.

Reply:
I'm thinking about exhaustion of the shared cpu pool and its implications. With this change, can we ever reach a point on which there are no cpus left in the shared pool?

Adding `minSharedCPUs` (final name TBD) will prevent that, effectively partitioning the CPUs in the node in 3 pools. But the existence of that third pool need to be exposed to enable the scheduler (and the system in general) to make the correct decisions. If nothing else, the CPUs available for exclusive allocation won't be `capacity`-`reserved` as it is today, but will be `capacity`-`reserved`-`minSharedCPUs`.

IOW, I don't see (yet?) a way to implement robust system behavior without somehow exposing to the rest of the system the `minSharedCPUs` pool size existence, which justifies the scheduler changes.

Reply:
> thank you for replying and confirming it is a valid option to keep the feature off forever (Alpha/Beta/GA).

Since this is not a norm (typically Beta onwards a feature is enabled by default), I would suggest explicitly discussing this during PRR (Performance Readiness Review) stage.

> How do I invite scheduler maintainers for comments? I did "/add-sig scheduling". Please advise.

I can see that SIG Scheduling is already cc'd on the KEP, they should triage it soon. If we don't hear from them in a week or so, please give a nudge on slack (https://kubernetes.slack.com/archives/C09TP78DV).